### ISSUE TITLE
[RFC] moving the observation system to synchronous timing

### ISSUE THREAD

**Issue by bigopon** [aurelia/aurelia#2035]
With glitch problem handled in #2020, we now only have state tearing issue left to solve for the timing of the observation system in v2. As noted in #1957, state tearing is an issue where there is a group of state changes that are supposed to be notified together, but are prematurely notified separately for individual state in it.
The state tearing causes some issues migrating some v1 libraries to v2.

---

_Originally posted by @3cp in https://github.com/aurelia/aurelia/issues/2033#issuecomment-2290303895_

              Like this:
```ts
class VM {
  service: SomeService;
  
  // Au1 will get true all the time
  // Au2 will have a false spike between a and b assignments.
  get computedState() {
    return this.service.a === this.service.b;
  }
}

class SomeService {
  a = 1;
  b = 1;

  someMethod() {
    this.a = 2;
    // Au2 introduced side effects between this two lines.
    // This is way against intuition.
    //
    // Note "SomeService" can be 3rd party lib that
    // User has no control of.
    //
    // I don't think this is a good practice.
    this.b = 2;
  }
}
```
            
From the above example, we can see there's a potential issue with synchronous system when used with 3rd party libraries. This is called state tearing as noted in #1957.
In this case, `someMethod()`, when called, will trigger the computed getter without being aware of that.

---

There are a few ways we can tackle this issue:

1. One way to fix this without changing the current timing of Aurelia is to have the ability to express that computed getter should only re-compute the next tick:
    ```ts
    class Vm {
    
      @flush({ async: true })
      get computedState() {
        return this.service.a === this.service.b
      }
    }
    ```
    `@flush({ ... })` decorator allows applications to be in control to deal with problematic third party code handling. Though with this approach, applications need to do it everywhere/every time that there's a reference to such code.

2. another way to fix this is to change the nature of aurelia observation system from synchronous to asynchronous. This is similar like how v1 behave, and state tearing will be automatically mitigated.

3. another way is to provide a way for 3rd party lib to retrieve the `batch()` function from Aurelia and call it in the mutation code where they are modifying multiple public properties at once. This is generally not how we want to write code, however. For example, with the above example, it'll be something like the following:
    ```ts
    export class SomeService {
      static inject = ['Au:batch']
      constructor(fn) {
        this.batch = fn;
      }

      someMethod() {
        this.batch(() => {
          this.a = 2;
          this.b = 2;
        });
      }
    }
    ```

---

One big question is what would the impact is on existing v2 application, if we are to move the observation timing back to asynchronous mode.
One way of mitigating the immediate issue is to have the ability to force changes to be immediately flushed, like a `noBatch` function, as opposed to `batch`:
```ts
class SomeService {
  someMethod() {
    noBatch(() => {
      this.a = 2;
      this.b = 2;
    });
  }
}
```

**Comment by bigopon** [comment]
We are wrapping for RC release, so please help discuss and give some thoughts on the matter if your team happens to rely on the synchronous timing instead of the reactive callback 🙏

**Comment by avrahamcool** [comment]
Although I can see the benefits of having all the bindings reflected synchronously, It's a big change from V1.
And that is a big change that may affect porting code to V2.

There are benefits to having async bindings.
This allows for a function to change a value multiple times, and to know that the binding will fire only once (at the end) with the relevant value.
For example, if a value is bounded to `x.y` I can easily see a function doing `x = {};` and then later on `x.y = 'abc';`
Or values that change based on some logic within the function (maybe a string that is being constructed with multiple `+=` operators).
I know that the examples that I'm giving here may not reflect good code, but those are things that may happen in a real-life application.
+ the example you gave of evaluating functions in between updates that may lead to a broken state.

Maybe the best way is to let the developer choose what kind of binding he wants.
(The default can be 'async' - just like V1, but a developer can opt in for sync binding when configuring the Aurelia project)
Using the `batch` or `noBatch` directives,  should override the default Aurelia configuration anyway.
That way: I can see a way to gradually transform code that was written for V1 to V2 without fear of major issues.

The `batch` and `noBatch` functions wrappers seems clunky to me.
Maybe they would be better off as decorators on a function?
Although the function syntax that you suggest is far more superior - because you can break the flow of a function into multiple parts, some `batch` and some `noBatch`.
That will be feasible but uglier with the decorator approach - because that will force the developer to break the code into multiple functions. (Maybe it's still better, though?)

So maybe Aurelia should also allow decorators for the cases where a complete function can be `batch`/`noBatch`?

You should specify what the rule is in case of nesting `batch` inside a `noBatch` and vice versa,
(or: using the `batch` function inside a code that was decorated with `noBatch`)
I think that the 'more nested' directive should apply.

I'm not even sure if all of what I'm suggesting here is even possible :)

**Comment by bigopon** [comment]
Thanks @avrahamcool , I think your feedback is valid. And somehow that helped me think of a mode where we have immediate flush on normal properties but async on computed properties + bindings. It feels like a good balanced as computed properties and bindings are the parts that "communicate"/"bridge" different parts of an application together.

For "batch"/"noBatch` juggling, I'm not sure if we want to do that much, if we have a good enough system and there' still some issues with timing, applications probably should just wait for a tick. For your other concerns related to implementation difficulty, I'm not sure, it may or may not be difficult.

**Comment by elitemike** [comment]
I haven't started my migration to V2 yet, but I personally think option 1 is the best.  It's nice and clean.  My only issue is the use of the flush decorator isn't self-explanatory to someone not familiar with it (ensure it's well documented).  Options are good.

Would it be possible to have a global setting that makes all computed getters behave async or synchronous?

**Comment by tareqimbasher** [comment]
Immediate flush on normal properties and async on computed ones sounds like a good default. And then pair that with:
- A decorator that can be used to override specific props
- An app-level config to change behavior globally instead of having to declare the decorator everywhere. This would enable backward-compat for existing Au2 apps that want to stay with the behavior they've built their app around

**Comment by 3cp** [comment]
I assume `<span if.bind="service.a === service.b">xxx</span>` is observing a computed property?

I agree default behaviour should aim for maximum compatibility and least surprise, not aim for performance. Then allow users to override it when they knew what they are doing.

**Comment by ekzobrain** [comment]
I agree with a mixed approach: sync mode for normal properties and async (but configurable) mode for computed getters and bindables. Example syntax:
```ts
// flushes async by default
get prop() {}

// enable sync flush, the behavior then would be as current
@computed({ sync: true })
get prop() {}
```

But I am against app-level settings because those may break plugins. A plugin author could not know what global settings might be in the target application. Generally making getters and bindables async should be safe enough for current V2 applications. I really like current sync binding system, except for some rare cases with state tearing. But I also agree that behaviour should be predictable.
